gui/app.py

```diff
@@ -20,32 +20,111 @@ class VideoThread(QThread):
 
     def __init__(self, filenames):
         super().__init__()
-        self._run_flag = True
+        self._thread_running_flag = True
+        self._video_playing_flag = True
         self._filenames = filenames
         self._captures = []
+        self._rewind = False
+
+    def __emit_frames(self):
+        """Emit next/prev frames on the pyqtSignal to be recieved by video widgets"""
+        for index, capture in enumerate(self._captures):
+            if (self._rewind):
+                prev_frame = cur_frame = capture.get(cv2.CAP_PROP_POS_FRAMES)
+                
+                if (cur_frame >= 2):
+                    # Go back 2 frames so when we read() we'll read back 1 frame
+                    prev_frame -= 2
+                else:
+                    # If at beginning, just read 1st frame over and over
+                    prev_frame = 0
+
+                capture.set(cv2.CAP_PROP_POS_FRAMES, prev_frame)
+            
+            # Read the frame
+            ret, cv_img = capture.read()
+            if ret:
+                self.update_frames_signal.emit(Frame(cv_img, index))
 
     def run(self):
         # Create list of video capturers
         for filename in self._filenames:
             self._captures.append(cv2.VideoCapture(filename))
 
-        # Emit Frames of the captures
-        while self._run_flag:
-            for index, capture in enumerate(self._captures):
-                ret, cv_img = capture.read()
-                if ret:
-                    self.update_frames_signal.emit(Frame(cv_img, index))
-            self.msleep(int(1000 / 30))
+        # Run the play/pausable video 
+        while self._thread_running_flag:
+            # Send frames while the video is playing
+            while self._thread_running_flag and self._video_playing_flag:
+                self.__emit_frames()
+                self.msleep(int(1000 / 30))
+            
+            # Take a NOP while its paused
+            while self._thread_running_flag and not self._video_playing_flag:
```

Comment:
You could get rid of the nested while loops to simplify things a bit.
```suggestion
            if self._video_playing_flag:
                self._emit_frames()
            
            self.msleep(...)
```